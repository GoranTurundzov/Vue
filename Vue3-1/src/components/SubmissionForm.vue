<template>
   <div class="content content-boxed bg-light">
        <div class="push-left">
            <button 
                class="btn btn-gray push-7-l"
                data-testid="go-back-form-button" 
                @click="closeForm"
            >
                <i class="fas fa-angle-left push-10-r"></i>Back
            </button>
        </div>
        <div class="justify-content-center">
            <h3>{{ title }}</h3>
        </div>
        <div class="content-narrow">
            <form class="form-horisontal block-content content" @submit="submitForm">
                <div class="row">
                    <div class="form-group" :class="{ 'has-error': !validName }">
                        <div class="col-xs-8 ">
                            <label for="Name" class="form-label floating open">Name*</label>
                        </div>
                        <div class="col-xs-12">
                            <input
                                v-model="submission.name"
                                type="text"
                                name="Name"
                                class="form-control"
                                @focus="removeNameError"
                                data-testid="name-input"
                            />
                        </div>
                    </div>
                    <div class="es-required-field col-md-6" v-if="!validName">
                        <i class="fas fa-exclamation-circle"></i>
                        This field is required.
                    </div>
                </div>
                <div class="row">
                    <div class="form-group" :class="{ 'has-error': emailError }">
                        <div class="col-xs-8">
                            <label for="Email" class="form-label floating open">Email*</label>
                        </div>
                        <div class="col-xs-12">
                            <input
                                v-model="submission.email"
                                type="text"
                                name="Email"
                                class="form-control"
                                @focus="removeEmailError"
                                data-testid="email-input"
                            />
                        </div>
                    </div>
                    <div class="es-required-field col-md-6 " v-if="!validEmail">
                        <i class="fas fa-exclamation-circle"></i>
                        This field is required.
                    </div>
                    <div class="es-required-field col-md-6 " v-if="!validEmailFormat">
                        <i class="fas fa-exclamation-circle"></i>
                        Enter a valid e-mail
                    </div>
                </div>
                <div class="row">
                    <div class="form-group">
                        <div class="col-xs-8">
                            <label for="Recommendation" class="form-label floating open">
                                Recommendation
                            </label>
                        </div>
                        <div class="col-xs-12">
                            <select
                                name="Recommendation"
                                v-model="submission.recommendation"
                                class="form-control"
                                data-testid="recommendation-input"
                            >
                                <option value="yes">Yes</option>
                                <option value="no">No</option>
                                <option value="maybe">Maybe</option>
                            </select>
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="form-group" :class="{ 'has-error': !validFeedback }">
                        <div class="col-xs-8">
                            <label for="Feedback" class="form-label floating open">Feedback*</label>
                        </div>
                        <div class="col-xs-12">
                            <textarea
                                name="Feedback"
                                v-model="submission.feedback"
                                rows="5"
                                class="form-control"
                                @focus="removeFeedbackError"
                                data-testid="feedback-input"
                            ></textarea>
                        </div>
                    </div>
                    <div class="es-required-field col-md-6" v-if="!validFeedback">
                        <i class="fas fa-exclamation-circle"></i>
                        This field is required.
                    </div>
                </div>
                <div class="row">
                    <div class="form-group" :class="{ 'has-error': dateError }">
                        <div class="col-xs-8">
                            <label for="Date" class="form-label floating open" type="date"
                                >Date*</label
                            >
                        </div>
                        <div class="col-xs-12">
                            <input
                                v-model="submission.date"
                                class="form-control"
                                type="text"
                                name="Date"
                                placeholder="DD/MM/YYYY"
                                @focus="removeDateError"
                                data-testid="date-input"
                            />
                        </div>
                    </div>
                    <div class="es-required-field col-md-6" v-if="dateError">
                        <div v-if="!validDate">
                            <i class="fas fa-exclamation-circle"></i>
                            This field is required.
                        </div>
                        <div v-if="!validDateFormat">
                            <i class="fas fa-exclamation-circle"></i>
                            Enter a valid date in dd/MM/yyyy format only.
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="form-group" :class="{ 'has-error': !validForm }">
                        <div class="col-xs-8">
                            <label for="Form" class="form-label floating open">Form*</label>
                        </div>
                        <div class="col-xs-12">
                            <select
                                v-model="submission.form"
                                name="Form"
                                class="form-control"
                                @change="removeFormError"
                                data-testid="form-input"
                            >
                                <option disabled selected value="">Select a form</option>
                                <option
                                    v-for="option of formSelectOptions"
                                    :value="option"
                                    :key="option.name"
                                    >{{ option.name }} - {{ option.form_type }}</option
                                >
                            </select>
                        </div>
                    </div>
                    <div class="es-required-field col-md-6" v-if="!validForm">
                        <i class="fas fa-exclamation-circle"></i>
                        This option is required.
                    </div>
                </div>
                <div class="row">
                    <div class="form-group">
                        <div class="col-xs-8">
                            <label for="Status" class="form-label floating open">Status</label>
                        </div>
                        <div class="col-xs-12">
                            <select
                                name="Status"
                                v-model="submission.status"
                                class="form-control"
                                data-testid="status-input"
                            >
                                <option value="pending">Pending</option>
                                <option value="approved">Approved</option>
                                <option value="rejected">Rejected</option>
                            </select>
                        </div>
                    </div>
                </div>

                <div class="form-group row">
                    <div class="col-xs-12 text-center">
                        <button
                            type="submit"
                            class="btn btn-primary form-material"
                            data-testid="submit-submission-button"
                        >
                            Submit
                        </button>
                    </div>
                </div>
            </form>
        </div>
    </div>
</template>

<script>
export default {
    props: {
        submissionToEdit: {
            type: Object,
            description: 'prop which will be passed to fill out the form with the submission to edit'
        },
        formSelectOptions: {
            type: Array,
            required: true,
            description: 'Options for the form select'
        },
        editedSubmissionIndex: {
            type: Number,
            default: -1,
            description: 'if this is passed the form will be used for editing'
        }
    },
    emits: ['close-form', 'create-new-submission', 'edit-the-submission'],
    data() {
        return {
            submission: {
                name: '',
                email: '',
                recommendation: 'yes',
                feedback: '',
                date: '',
                form: '',
                status: 'pending'
            },
            validName: true,
            validEmail: true,
            validFeedback: true,
            validForm: true,
            validDate: true,
            validDateFormat: true,
            validEmailFormat: true
        };
    },
    computed: {
        dateError() {
            if (!this.validDate || !this.validDateFormat) {
                return true;
            }
            return false;
        },
        emailError() {
            if (!this.validEmail || !this.validEmailFormat) {
                return true;
            }
            return false;
        }
    },
    created() {
        if (this.editedSubmissionIndex >= 0) {
            this.title = 'Edit ' + this.submissionToEdit.name + ' Submission';
            this.submission.name = this.submissionToEdit.name;
            this.submission.email = this.submissionToEdit.email;
            this.submission.recommendation = this.submissionToEdit.recommendation;
            this.submission.feedback = this.submissionToEdit.feedback;
            this.submission.form = {
                name: this.submissionToEdit.form,
                form_type: this.submissionToEdit.form_type
            };
            this.submission.status = this.submissionToEdit.status;
            this.submission.date = this.submissionToEdit.date;
        } else {
            this.title = 'Create new submission';
        }
    },
    methods: {
        submitForm(e) {
            e.preventDefault();
            this.validName = this.submission.name.trim() === '' ? false : true;
            this.validEmail = this.submission.email.trim() === '' ? false : true;
            this.validFeedback = this.submission.feedback.trim() === '' ? false : true;
            this.validForm = this.submission.form === '' ? false : true;
            this.validDate = this.submission.date.trim() === '' ? false : true;
            if (this.validDate) {
                this.validDateFormat = this.dateValidation();
            }
            if (this.validEmail) {
                this.validEmailFormat = this.emailValidation();
            }
            if (
                this.validName &&
                this.validEmail &&
                this.validFeedback &&
                this.validForm &&
                this.validDate &&
                this.validEmailFormat &&
                this.validDateFormat
            ) {
                if (this.editedSubmissionIndex < 0) {
                    this.$emit('create-new-submission', this.submission);
                } else {
                    this.$emit('edit-the-submission', this.submission, this.editedSubmissionIndex);
                }
            }
        },
        closeForm() {
            this.$emit('close-form');
        },
        dateValidation() {
            var reg = /(0[1-9]|[12][0-9]|3[01])[- /.](0[1-9]|1[012])[- /.](19|20)\d\d/;
            if (this.submission.date.match(reg)) {
                let parts = this.submission.date.split('/');
                var enteredTime = new Date(parts[1] + '/' + parts[0] + '/' + parts[2]);
                var timeNow = new Date();
                if (enteredTime <= timeNow) {
                    return true;
                }
            }
            return false;
        },
        removeNameError() {
            this.validName = true;
        },
        removeDateError() {
            this.validDate = true;
            this.validDateFormat = true;
        },
        removeEmailError() {
            this.validEmail = true;
            this.validEmailFormat = true;
        },
        removeFormError() {
            this.validForm = true;
        },
        removeFeedbackError() {
            this.validFeedback = true;
        },
        emailValidation() {
            const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
            return re.test(String(this.submission.email).toLowerCase());
        }
    }
};
</script>

<style scoped>
textarea {
    resize: none;
}
.es-required-field {
    color: #ff3636;
    display: block;
    font-size: 14px;
    margin-top: 8px;
    text-align: left;
}
</style>
