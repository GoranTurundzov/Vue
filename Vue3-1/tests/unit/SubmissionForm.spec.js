import { mount } from '@vue/test-utils';
import SubmissionForm from '../../src/components/SubmissionForm';
import formJson from '../../public/forms.json';
import dummyData from '../../src/assets/DummyData';

let wrapper;

let spySubmitFormFn = jest.spyOn(SubmissionForm.methods, 'submitForm');
let spyEmailValidationFn = jest.spyOn(SubmissionForm.methods, 'emailValidation');
let spyDateValidationFn = jest.spyOn(SubmissionForm.methods, 'dateValidation');
let spyCloseFormFn = jest.spyOn(SubmissionForm.methods, 'closeForm');

describe('SubmissionForm', () => {
    beforeEach(() => {
        wrapper = mount(SubmissionForm, {
            props: {
                formSelectOptions: formJson,
                editedSubmissionIndex: -1
            }
        });
        wrapper.setData({
            submission: {
                name: '',
                email: '',
                recommendation: 'yes',
                feedback: '',
                date: '',
                form: '',
                status: 'pending'
            },
            validName: true,
            validEmail: true,
            validFeedback: true,
            validForm: true,
            validDate: true,
            validDateFormat: true,
            validEmailFormat: true
        });
    });

    it('App renders', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    test('Date Validatot works', async () => {
        let dateInput = wrapper.find('[data-testid="date-input"]');

        expect(wrapper.vm.validDateFormat).toBe(true);
        expect(wrapper.vm.validDate).toBe(true);

        dateInput.element.value = '';
        dateInput.trigger('input');

        await wrapper.find('form').trigger('submit');
        expect(wrapper.vm.validDateFormat).toBe(true);
        expect(wrapper.vm.validDate).toBe(false);

        dateInput.element.value = 'asddsa';
        dateInput.trigger('input');
        await wrapper.find('form').trigger('submit');
        expect(spyDateValidationFn).toBeCalled();
        expect(wrapper.vm.validDateFormat).toBe(false);
        expect(wrapper.vm.validDate).toBe(true);

        dateInput.element.value = '11/11/2020';
        dateInput.trigger('input');
        await wrapper.find('form').trigger('submit');
        expect(spyDateValidationFn).toBeCalled();
        expect(wrapper.vm.validDateFormat).toBe(true);
        expect(wrapper.vm.validDate).toBe(true);
    });

    test('Email validator works', async () => {
        let emailInput = wrapper.find('[data-testid="email-input"]');

        await wrapper.find('form').trigger('submit');
        expect(wrapper.vm.validEmail).toBe(false);
        expect(wrapper.vm.validEmailFormat).toBe(true);

        emailInput.element.value = 'Goran';
        emailInput.trigger('input');
        await wrapper.find('form').trigger('submit');
        expect(spyEmailValidationFn).toBeCalled();
        expect(wrapper.vm.validEmail).toBe(true);
        expect(wrapper.vm.validEmailFormat).toBe(false);

        emailInput.element.value = 'Goran@gmail.com';
        emailInput.trigger('input');
        await wrapper.find('form').trigger('submit');
        expect(spyEmailValidationFn).toBeCalled();
        expect(wrapper.vm.validEmail).toBe(true);
        expect(wrapper.vm.validEmailFormat).toBe(true);
    });

    it('Renders the title', async () => {
        await wrapper.setData({
            title: 'Submission Creation'
        });
        expect(wrapper.find('h3').text()).toEqual(wrapper.vm.title);
    });

    test('Form validation and submission works', async () => {
        let nameInput = wrapper.find('[data-testid="name-input"]');
        let emailInput = wrapper.find('[data-testid="email-input"]');
        let recommendationInput = wrapper.find('[data-testid="recommendation-input"]');
        let statusInput = wrapper.find('[data-testid="status-input"]');
        let feedbackInput = wrapper.find('[data-testid="feedback-input"]');
        let formInput = wrapper.find('[data-testid="form-input"]');
        let dateInput = wrapper.find('[data-testid="date-input"]');

        //inserting invalid inputs
        nameInput.element.value = '';
        nameInput.trigger('input');
        emailInput.element.value = 'Goran@gmail.com';
        emailInput.trigger('input');
        feedbackInput.element.value = 'the feedback string';
        feedbackInput.trigger('input');
        dateInput.element.value = 'asdas0';
        dateInput.trigger('input');
        recommendationInput.findAll('option')[1].setSelected();
        statusInput.findAll('option')[1].setSelected();
        formInput.findAll('option')[4].setSelected();

        await wrapper.find('form').trigger('submit');

        expect(wrapper.vm.validName).toBe(false);
        expect(wrapper.vm.validDateFormat).toBe(false);
        expect(spySubmitFormFn).toBeCalled();
        expect(spyEmailValidationFn).toBeCalled();
        expect(spyDateValidationFn).toBeCalled();
        expect(wrapper.emitted('create-new-submission', wrapper.vm.submission)).toBeFalsy();

        //inserting valid inputs
        dateInput.element.value = '11/11/2020';
        dateInput.trigger('input');
        nameInput.element.value = 'Goran';
        nameInput.trigger('input');

        await wrapper.find('form').trigger('submit');

        expect(wrapper.vm.validName).toBe(true);
        expect(wrapper.vm.validDate).toBe(true);
        expect(spySubmitFormFn).toBeCalled();
        expect(wrapper.emitted('create-new-submission', wrapper.vm.submission)).toBeTruthy();
        expect(wrapper.emitted('create-new-submission', wrapper.vm.submission)).toHaveLength(1);
        expect(wrapper.emitted('create-new-submission')[0][0]).toEqual(wrapper.vm.submission);
    });

    test('Back button', async () => {
        let backButton = wrapper.find('[data-testid="go-back-form-button"]');
        await backButton.trigger('click');
        expect(spyCloseFormFn).toBeCalled();
        expect(wrapper.emitted('close-form')).toBeTruthy();
        expect(wrapper.emitted('close-form')).toHaveLength(1);
    });

    test('Error messages appear when invalid input', async () => {
        let emailInput = wrapper.find('[data-testid="email-input"]');
        let recommendationInput = wrapper.find('[data-testid="recommendation-input"]');
        let statusInput = wrapper.find('[data-testid="status-input"]');
        let feedbackInput = wrapper.find('[data-testid="feedback-input"]');
        let formInput = wrapper.find('[data-testid="form-input"]');
        let dateInput = wrapper.find('[data-testid="date-input"]');

        expect(wrapper.findAll('.es-required-field')).toHaveLength(0);

        emailInput.element.value = 'Goran@gmail.com';
        emailInput.trigger('input');
        feedbackInput.element.value = 'the feedback string';
        feedbackInput.trigger('input');
        dateInput.element.value = 'asdas0';
        dateInput.trigger('input');
        recommendationInput.findAll('option')[1].setSelected();
        statusInput.findAll('option')[1].setSelected();
        formInput.findAll('option')[4].setSelected();

        await wrapper.find('form').trigger('submit');
        expect(spySubmitFormFn).toBeCalled();
        //it is expected to have 2 because the date format is invalid and there is nothing in the name field
        expect(wrapper.findAll('.es-required-field')).toHaveLength(2);
    });

    test('If an index is passed it loads that submission for editing and submits it', async () => {
        let editWrapper = mount(SubmissionForm, {
            props: {
                formSelectOptions: formJson,
                editedSubmissionIndex: 2,
                submissionToEdit: dummyData[2]
            }
        });
        let nameInput = editWrapper.find('[data-testid="name-input"]');
        let emailInput = editWrapper.find('[data-testid="email-input"]');
        let recommendationInput = editWrapper.find('[data-testid="recommendation-input"]');
        let statusInput = editWrapper.find('[data-testid="status-input"]');
        let feedbackInput = editWrapper.find('[data-testid="feedback-input"]');
        let formInput = editWrapper.find('[data-testid="form-input"]');
        let dateInput = editWrapper.find('[data-testid="date-input"]');

        expect(editWrapper.vm.submission).toEqual({
            name: 'Jonathan',
            email: 'jonathan@gmail.com',
            recommendation: 'maybe',
            feedback: 'My neighbor Karly has one of these. She works as a gambler and she says it looks tall. I saw one of these in Spratly Islands and I bought one. works okay.',
            date: '16/03/2021',
            form: { name: 'Form 3', form_type: 'page' },
            status: 'approved'
        });

        nameInput.element.value = 'Not Jonathan';
        nameInput.trigger('input');
        emailInput.element.value = 'Nojonathan@gmail.com';
        emailInput.trigger('input');
        feedbackInput.element.value = 'Changed';
        feedbackInput.trigger('input');
        dateInput.element.value = '16/07/2021';
        dateInput.trigger('input');
        recommendationInput.findAll('option')[2].setSelected();
        statusInput.findAll('option')[2].setSelected();
        formInput.findAll('option')[2].setSelected();

        expect(editWrapper.vm.submission).toEqual({
            name: 'Not Jonathan',
            email: 'Nojonathan@gmail.com',
            recommendation: 'maybe',
            feedback: 'Changed',
            date: '16/07/2021',
            form: { id: 24384, name: 'Form 2', form_type: 'page' },
            status: 'rejected'
        });

        await editWrapper.find('form').trigger('submit');
        expect(spySubmitFormFn).toBeCalled();
        expect(editWrapper.emitted('edit-the-submission', editWrapper.vm.submission, editWrapper.vm.submissionToEdit)).toBeTruthy();
        expect(editWrapper.emitted('edit-the-submission')).toHaveLength(1);
        expect(editWrapper.emitted('edit-the-submission')[0][0]).toEqual(editWrapper.vm.submission, editWrapper.vm.submissionToEdit);
    });
});
