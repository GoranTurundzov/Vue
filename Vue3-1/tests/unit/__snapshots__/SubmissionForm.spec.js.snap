// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionForm App renders 1`] = `
<div
  class="content content-boxed bg-light"
>
  <div
    class="push-left"
  >
    <button
      class="btn btn-gray push-7-l"
      data-testid="go-back-form-button"
    >
      <i
        class="fas fa-angle-left push-10-r"
      />
      Back 
    </button>
  </div>
  <div
    class="justify-content-center"
  >
    <h3>
      Create new submission
    </h3>
  </div>
  <div
    class="content-narrow"
  >
    <form
      class="form-horisontal block-content content"
    >
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8 "
          >
            <label
              class="form-label floating open"
              for="Name"
            >
              Name*
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <input
              class="form-control"
              data-testid="name-input"
              name="Name"
              type="text"
            />
          </div>
        </div>
        <!--v-if-->
      </div>
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8"
          >
            <label
              class="form-label floating open"
              for="Email"
            >
              Email*
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <input
              class="form-control"
              data-testid="email-input"
              name="Email"
              type="text"
            />
          </div>
        </div>
        <!--v-if-->
        <!--v-if-->
      </div>
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8"
          >
            <label
              class="form-label floating open"
              for="Recommendation"
            >
               Recommendation 
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <select
              class="form-control"
              data-testid="recommendation-input"
              name="Recommendation"
            >
              <option
                value="yes"
              >
                Yes
              </option>
              <option
                value="no"
              >
                No
              </option>
              <option
                value="maybe"
              >
                Maybe
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8"
          >
            <label
              class="form-label floating open"
              for="Feedback"
            >
              Feedback*
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <textarea
              class="form-control"
              data-testid="feedback-input"
              name="Feedback"
              rows="5"
            />
          </div>
        </div>
        <!--v-if-->
      </div>
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8"
          >
            <label
              class="form-label floating open"
              for="Date"
              type="date"
            >
              Date*
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <input
              class="form-control"
              data-testid="date-input"
              name="Date"
              placeholder="DD/MM/YYYY"
              type="text"
            />
          </div>
        </div>
        <!--v-if-->
      </div>
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8"
          >
            <label
              class="form-label floating open"
              for="Form"
            >
              Form*
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <select
              class="form-control"
              data-testid="form-input"
              name="Form"
            >
              <option
                disabled=""
                value=""
              >
                Select a form
              </option>
              
              <option
                value="[object Object]"
              >
                Form 1 - popup
              </option>
              <option
                value="[object Object]"
              >
                Form 2 - page
              </option>
              <option
                value="[object Object]"
              >
                Form 3 - page
              </option>
              <option
                value="[object Object]"
              >
                Form 4 - popup
              </option>
              <option
                value="[object Object]"
              >
                Form 5 - email
              </option>
              <option
                value="[object Object]"
              >
                Form 6 - page
              </option>
              <option
                value="[object Object]"
              >
                Form 7 - email
              </option>
              
            </select>
          </div>
        </div>
        <!--v-if-->
      </div>
      <div
        class="row"
      >
        <div
          class="form-group"
        >
          <div
            class="col-xs-8"
          >
            <label
              class="form-label floating open"
              for="Status"
            >
              Status
            </label>
          </div>
          <div
            class="col-xs-12"
          >
            <select
              class="form-control"
              data-testid="status-input"
              name="Status"
            >
              <option
                value="pending"
              >
                Pending
              </option>
              <option
                value="approved"
              >
                Approved
              </option>
              <option
                value="rejected"
              >
                Rejected
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-xs-12 text-center"
        >
          <button
            class="btn btn-primary form-material"
            data-testid="submit-submission-button"
            type="submit"
          >
             Submit 
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
